designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+assign an `int` expression to a `real` variable but not vice-versa.
+Covariant typing would extend this as follows.
+
+* `int` is a subtype `real`,
+* `T[]` is a subtype of `U[]` if `T` is a subtype of `U`, and
+* `T0(T1, ..., TN)` is a subtype of `U0(U1, ..., UN)` if
+  `T0` is a subtype of `U0`, `U1` is a subtype of `T1`, ..., `UN` is a
+  subtype of `TN`.
+
+The type `T` in the array type `T[]` is covariant in that it preserves
+subtyping.  The type `T` in `T(U)` is similarly covariant, but `U` is
+contravariant, in that it reverses the subtyping relationship.  As a
+concrete example, consider legal types for `a` (the lvalue) and `b`
+(the rvalue) in `a = b`,
+
+| lvalue type  | legal rvalue types                     |
+|:------------:|:--------------------------------------:|
+| `int(int)`   | `int(int)`, `int(real)`                |
+| `real(real)` | `real(real)`, `int(real)`              |
+| `int(real)`  | `int(real)`                            |
+| `real(int)`  | `real(int)`, `real(real)`, `int(real)` |
+
+
+#### Sized types
+
+The set of sized runtime types is the least set of types such that
+
+* *sized primitive type*: `int` and `real` are sized types,
+* *sized vector type*: `vector[K]` and `row_vector[K]` are sized
+types if `K` is a non-negative integer,
+* *sized matrix type*: `matrix[M, N]` is a sized type if `M` and `N` are
+  non-negative integers,
+* *sized array type*: `T[]` is a sized type if `T` is a sized type, and
+* *sized function type*: `T0(T1,...,TN)` is a sized type if `T0`, `T1`, ..., `TN`
+are sized types.
+
+
+## Lambda Expression Syntax
+
+Syntactically, a lambda expression is represented as a function argument
+list followed by a function body.   For example, in `(real u) { return
+1 / (1 + exp(-u)); }`, the function argument list is `(real u)` and
+the function body is `{ return 1 / (1 + exp(-u)); }`.
+
+As a shorthand, a function argument list followed by an expression is
+taken to be shorthand for the function argument list followed by a
+function body returning that expression.  For example, the lambda
+expression
+
+```
+(real u) { return 1 / (1 + exp(-u)); }
+```
+
+may be replaced with the equivalent
+
+```
+(real u) 1 / 1 + exp(-u)
+```
+
+Lambda expressions behave like other expressions such as `2 + 2`, only
+they denote functions rather than values and have function types.
+
+
+## Changes to Standard Functions
+
+Allow functions to be defined in any scope.  Allow function
+definitions and lambda expressions to
+capture variables in scope by references.  
+
+Deprecate the functions block;  functions declared there can be
+declared in the transformed data block.  
+
+Standard function definitions of the form
+
+```
+T0 f(T1 x1, ..., TN xN) { ... }
+```
+
+may be replaced with
+
+```
+T0(T1, ..., TN) f = (T1 x1, ..., TN xN) { ... }
+```
+
+## Variable declarations
+
+Variable declarations for functions follows the usual syntax, with a
+sized type required for block variables, an unsized type for function
```

Comment:
Does this mean you need a "sized function type" if you allow functions at the top level? What does that buy you over an unsized function type?